code/controllers/subsystem/ipintel.dm

```diff
@@ -1,13 +1,273 @@
+	if(intel.result >= check_probability)
+		return IPINTEL_BAD_IP
+	return IPINTEL_GOOD_IP
+
+/datum/controller/subsystem/ipintel/proc/is_rate_limited()
+	var/static/minute_key
+	var/expected_minute_key = floor(REALTIMEOFDAY / 1 MINUTES)
+
+	if(minute_key != expected_minute_key)
+		minute_key = expected_minute_key
+		rate_limits[IPINTEL_RATE_LIMIT_MINUTE] = 0
+
+	if(rate_limits[IPINTEL_RATE_LIMIT_MINUTE] >= max_queries_per_minute)
+		return IPINTEL_RATE_LIMITED_MINUTE
+	if(rate_limits[IPINTEL_RATE_LIMIT_DAY] >= max_queries_per_day)
+		return IPINTEL_RATE_LIMITED_DAY
+	return FALSE
+
+/datum/controller/subsystem/ipintel/proc/query_address(address, allow_cached = TRUE)
+	if(allow_cached && fetch_cached_ip_intel(address))
+		return cached_queries[address]
+	var/is_rate_limited = is_rate_limited()
+	if(is_rate_limited)
+		return is_rate_limited
+	if(!initialized)
+		return IPINTEL_UNKNOWN_INTERNAL_ERROR
+
+	rate_limits[IPINTEL_RATE_LIMIT_MINUTE] += 1
+	rate_limits[IPINTEL_RATE_LIMIT_DAY] += 1
+
+	var/query_base = "https://[src.query_base]/check.php?ip="
+	var/query = "[query_base][address]&contact=[contact_email]&flags=b&format=json"
+
+	var/datum/http_request/request = new
+	request.prepare(RUSTG_HTTP_METHOD_GET, query)
+	request.execute_blocking()
+	var/datum/http_response/response = request.into_response()
+	var/list/data = response.body
+
+	var/datum/ip_intel/intel = new
+	intel.query_status = data["status"]
+	if(intel.query_status != "success")
+		return intel
+	intel.result = data["result"]
+	intel.date = SQLtime()
+	intel.address = address
+	cached_queries[address] = intel
+	add_intel_to_database(intel)
+	return intel
+
+/datum/controller/subsystem/ipintel/proc/add_intel_to_database(datum/ip_intel/intel)
+	var/datum/db_query/query = SSdbcore.NewQuery(
+		"INSERT INTO [format_table_name("ipintel")] ( \
+			ip, \
+			intel, \
+		) VALUES ( \
+			INET_ATON(:address) \
+			:result, \
+		)", intel.serialize_list(list(), list()))
+	query.warn_execute()
+	query.sync()
+	qdel(query)
+
+/datum/controller/subsystem/ipintel/proc/fetch_cached_ip_intel(address)
+	var/date_restrictor
+	if(ipintel_cache_length > 0)
+		date_restrictor = " AND date > DATE_SUB(NOW(), INTERVAL [ipintel_cache_length] DAY)"
+	var/datum/db_query/query = SSdbcore.NewQuery(
+		"SELECT * FROM [format_table_name("ipintel")] WHERE ip = INET_ATON(:address)[date_restrictor]", list(
+			"address" = address
+		)
+	)
+	query.warn_execute()
+	query.sync()
+	if(query.status == DB_QUERY_BROKEN)
+		qdel(query)
+		return null
+
+	query.NextRow()
+	var/list/data = query.item
+	qdel(query)
+	if(isnull(data))
+		return null
+
+	var/datum/ip_intel/intel = new
+	intel.query_status = "cached"
+	intel.result = data["intel"]
+	intel.date = data["date"]
+	intel.address = address
+	return TRUE
+
+/datum/controller/subsystem/ipintel/proc/is_exempt(client/player)
+	if(exempt_living_playtime > 0)
+		var/list/play_records = player.prefs.exp
+		if (!play_records.len)
+			player.set_exp_from_db()
```

Comment:
are we sure this wont set it to zero on their first connection?

Reply:
it only fetches from the db, doesn't set to it
(if I understood the code)

Reply:
ok i trust you

Reply:
`play_records[EXP_TYPE_LIVING]` would only become zero when the db has no row for their living playtime or if it's already zero. The proc won't modify any db value.

Reply:
Wouldn't that be a problem, if they have no living time they'll pass the zero minute check cause it defaults to zero

Reply:
no because we only check here if the config is set above zero .
I imagine it'll be set to something like five or ten minutes. Maybe even one